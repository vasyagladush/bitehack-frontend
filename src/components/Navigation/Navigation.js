import React,{useState} from 'react';
import classes from './Navigation.module.css';
import Backdrop from '../Backdrop/Backdrop';
import DrawerToggle from '../DrawerToggle/DrawerToggle';

const NavigationBar = props=>{
    const [show,setShow] = useState(false);

    const changeRoute = url =>{
        props.history.push(url);
    };

    const handleSidebar = ()=>{
        setShow(prevstate=>!prevstate);
    }

    return (
        <>
        <div className={classes.navDesktop}>
            {props.social ? <p className={classes.active}> 
                Social
            </p> : <p onClick={()=> changeRoute('/social')}>
                Social
            </p> // routing 
            }
            {props.home ? <p className={classes.active}>Home</p> : 
            <p onClick={()=> changeRoute('/')}>
                Home
            </p> }
            {props.about ? <p className={classes.active}>About</p> :
            <p onClick={()=> changeRoute('/about')}>
                About
            </p>
            }
        </div>
        <DrawerToggle clicked={handleSidebar}/>
            <ul className={show ? [classes.open,classes.mobile].join(' ') : [classes.close,classes.mobile].join(' ')}>
                <li>
                    <img src={require('../../images/messenger.png')} alt="messenger logo" width='50%' height='50%' style={{marginLeft: 10,marginBottom: 10}}/>
                </li>
                <li>{props.social ? <p className={classes.active}> 
                    Social
                </p> : <p onClick={()=> changeRoute('/social')}>
                    Social
                </p> // routing 
                }</li>
                <li>
                {props.home ? <p className={classes.active}>Home</p> : 
                <p onClick={()=> changeRoute('/')}>
                    Home
                </p> } 
                </li>
                <li>
                {props.about ? <p className={classes.active}>About</p> :
                <p onClick={()=> changeRoute('/about')}>
                    About
                </p>
                }
                </li>
            </ul>
        <Backdrop show={show} clicked={handleSidebar}/>
        </>
    )
}; 

export default NavigationBar;