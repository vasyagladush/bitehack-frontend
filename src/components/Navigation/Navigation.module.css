.navDesktop {
    display: flex;
    justify-content: left;
    width: 50%;
    margin-left: 20px;
}
.navDesktop p{
    margin: 10px 10% 10px 10%;
    font-weight: bold;
}

.navDesktop p:hover{
    color: #008c8c;
    cursor: pointer;
}

.mobile {
    position: absolute;
    list-style-type: none;
    top: 0px;
    z-index: 100;
    height: 100%;
    width: 30%;
    border-right: 1px solid;
    border-color: black;
    background-color: #FFFFFF;
    margin: auto;
    transition: transform 0.3s ease-out;
}

.mobile p{
    margin: 10% 10% 10% 10%;
    font-weight: bold;
}

.mobile p:hover{
    color: #008c8c;
    cursor: pointer;
}

.open {
    transform: translateX(0);
}

.close {
    transform: translateX(-100%);
}

.mobile li{
    font-weight: bold;
}


.active {
    color: #008c8c;
}

@media (max-width: 650px){
    .navDesktop {
        display: none;
    }
    .mobile{
        display: inline-block;
    }
}