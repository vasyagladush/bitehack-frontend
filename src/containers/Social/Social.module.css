.socialinput {
    width: 50%;
    margin: auto;
    margin-top: 50px;
}

.ul { 
    list-style-type: none;
    width: 600px;
    margin: auto;
}

.li { 
    border: 1px solid gainsboro;
    margin-top: 30px;
    border-radius: 10px;
    text-align: center;
    margin-right: 30px;
}

.Button {
    margin-bottom: 10px;
}

.Button_logout {
    position: absolute; 
    margin-left: 20px; 
    margin-top: 5px;
    top: 1px;
    width: 20%;
    height: 50px;
}

.not_find_user { 
    text-align: center;
    font-size: 20px;
    padding-right: 20%;
}

@media (max-width: 650px){ 
    .container {
        text-align: center;
        margin-right: 45%;
    }
    .Button_logout{
        width: 40%;
    }
    .li{
        width: 70%;
    }
}