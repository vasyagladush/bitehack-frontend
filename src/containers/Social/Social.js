import React, { useState, useEffect,useCallback } from 'react';
import Navigation from '../../components/Navigation/Navigation';
import Emoji from '../../components/Emoji/Emoji';
import {Form,Button,Spinner} from 'react-bootstrap';
import classes from './Social.module.css';
import axios from 'axios';
import * as actions from '../../store/actions/index';
import {connect} from 'react-redux';
import {url} from '../../ApiUrl';

const Social = props =>{
    const [state,setState] = useState({
        input: '',
        accounts: [],
        friends: [],
        IdOfMe: '',
        loading: false
    });
    const handleChange = event=>{
        setState({
            ...state,
            input: event.target.value
        });

    }

    const loginButtonHandler = useCallback(()=>{
        props.history.push('/');
    },[props]);

    const getUsersAndFriends = useCallback(() => {
        let accounts; // we must do it like this because this is not working if we want to setState
            axios.get(url+"/user/users",{headers: {
                "Authorization": "Bearer "+localStorage.getItem('token'),
                "input": state.input
            }}).then(response=>{
                accounts = response.data.accounts;
                axios.get(url+'/user/friends',{ // getting friends and users to filter it in social tab 
                    headers:{
                        "Authorization": 'Bearer '+ localStorage.getItem('token')
                    }
                }).then(response=>{
                    setState(prevState =>{return {
                        ...prevState,
                        friends: response.data.friends,
                        accounts: accounts,
                        loading: false
                    }});
                }).catch(err=>{
                    console.log(err);
                })
            });
    },[state.input]); 

    const getMe = useCallback(()=>{
        axios.get(url+'/user/me',{headers:{
            "Authorization": 'Bearer '+ localStorage.getItem('token')
        }}).then(response=>{
            props.setMe(response.data.id);
        })
        .catch(err=>{
            console.log(err);
        })
    },[props]);

    const addFriend = (id) =>{
        axios.post(url+'/user/add_friend',{
            id: id.toString()
        },{headers:{
            "Authorization": 'Bearer '+localStorage.getItem('token')
        }})
        .then(response=>{
            getUsersAndFriends();
            })
            .catch(err=>{
                console.log(err);
            });
    }

    const acceptFriend = (id)=>{
        axios.patch(url+'/user/accept_friend',{
            id: id.toString()
        },{headers:{
            'Authorization': 'Bearer '+localStorage.getItem('token')
        }}).then(response=>{
                getUsersAndFriends();
            })
            .catch(err=>{
                console.log(err);
            });
}

    const filterUsers = ()=>{ // filter user to not get yourself and displaying it actual status in relation to Account logged
        let temp = [];
        state.accounts.map(account=>{
            return state.friends.map(friend=>{
                if (account._id.toString() === friend.friend._id.toString())
                {
                 temp.push(account._id); // if we have friend 
                }
                return null;
            });
        });
        const filtered = state.accounts.filter(people=> people._id.toString() !== props.id.toString()); // filter to not get yourself 
        return filtered.map(people=>{
            let temp2 = false;
            let temp3 = null; 
            return (
                <li className={classes.li} key={people._id}>
                    <p>{people.name}</p>
                    <p>{people.surname}</p>
                    {
                        state.friends.map(friend=>{
                            if(people._id.toString()===friend.friend._id.toString())
                            {
                                if(friend.send===true&&friend.accepted===false)
                                {
                                    temp3 = (<p style={{fontWeight: 'bold'}}>Waiting for accept <Emoji symbol="👍🏻"/></p>);
                                    temp2 = true;
                                }
                                else if(friend.send===false&&friend.accepted===false)
                                {
                                    temp3 = (<Button variant="dark" className={classes.Button} onClick={()=>acceptFriend(people._id)}>Accept<Emoji symbol="🤝"/></Button>);
                                    temp2 = true;
                                }
                                else if(friend.accepted===true)
                                {
                                    temp3 = (<p style={{fontWeight: 'bold'}}>This is your friend <Emoji symbol="👨🏽‍💻"/></p>);
                                    temp2 = true;
                                }
                            }
                            return null;
                        })
                    }
                    {
                        temp2 ? temp3 : <Button variant="dark" className={classes.Button} onClick={()=>
                            addFriend(people._id)}>Add Friend<Emoji symbol="👋"/></Button>
                    } 

                </li>
            )
        });
    };

    useEffect(()=>{
        let timer;
        if (localStorage.token)
        {
            setState(prevState=>{return{
                ...prevState,
                loading: true
            }});
            timer = setTimeout(()=>{
                getUsersAndFriends();
            },400);
        }
        return ()=> clearTimeout(timer);
    },[getUsersAndFriends]);

    useEffect(()=>{
        if (localStorage.token)
        {
        getMe();
        }
    },[getMe]);

    const users = filterUsers();

    if (localStorage.token)
    {
    return (
        <>
        <Navigation social history={props.history}/>
        <div className={classes.container}>
            <Button variant="dark" className={classes.Button_logout} onClick={()=>{
                    localStorage.removeItem('token');
                    window.location.reload(false);
            }}>logout</Button>
        </div>
        <Form className={classes.socialinput}>
            <Form.Group controlId="formBasicEmail">
                <Form.Control type="text" placeholder="Search Friends by typing their names 🧙‍♂️" value={state.input} onChange={handleChange}/>
            </Form.Group>
        </Form>
        <ul className={classes.ul}>
            {
                users.length ? users : state.loading ? null : <h1 className={classes.not_find_user}>We can't find users <Emoji symbol="😵"/></h1>
            }
        </ul>
        <div style={{textAlign: 'center'}}>
        {
                state.loading ? <Spinner animation="border" style={{width: '3rem', height: '3rem',color: 'black'}}/> : null
        }
        </div>
    </>
    )
        }
        else {
            return(
                <>
                <Navigation social history={props.history}/>
                <h1 style={{marginTop: '9%',textAlign: 'center',fontSize: 50}}>Login first <Emoji symbol="📜"/></h1>
                <h1 style={{marginTop: '9%',textAlign: 'center',fontSize: 50}}>And Start Messaging <Emoji symbol="✏️"/> </h1>
                <div style={{textAlign: 'center'}}>
                    <Button variant="dark" style={{marginTop: '75px', width: '30%'}} onClick={loginButtonHandler}>Login</Button>
                </div>
                </>
            )
        }
};

const mapStateToProps = state=>{
    return {
        id: state.meId
    }
}
const mapDispatchToProps = dispatch=>{
    return {
        setMe: (id)=>dispatch(actions.addMe(id))
    }
};

export default connect(mapStateToProps,mapDispatchToProps)(Social);