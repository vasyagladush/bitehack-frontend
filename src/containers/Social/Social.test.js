import React from 'react';

import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';
import {configure,mount, shallow, render} from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import Navigation from '../../components/Navigation/Navigation';
import Social from './Social';
import {Spinner,Button} from 'react-bootstrap';

let store = configureMockStore()({});
configure({adapter: new Adapter});

describe('<Social/>',()=>{
    let wrapper;
    beforeEach(()=>{
        wrapper = shallow(
            <Provider store={store}>
                <Social />
            </Provider>
        );
    });
    it('matches snapshot',()=>{
        localStorage.setItem('token','eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVmM2MzYzYwZjkyZjNhMWUzYzc1ZjIwZiIsImlhdCI6MTU5ODM0ODEyMn0.lZFeZwwL0kevnI--7bKmY8tEX1W5Vwx2zedwvqneCws');
        wrapper = renderer.create(
            <Provider store={store}>
            <Social />
        </Provider>
        ).toJSON();
        expect(wrapper).toMatchSnapshot();
    });
    it('should render Navigation', ()=>{
        expect(wrapper.find(Navigation)).toBeTruthy();
    });
    it('should render spinner and input if we are logged',()=>{
        expect(wrapper.find('input')).toBeTruthy();
        expect(wrapper.find(Spinner)).toBeTruthy();
        expect(wrapper.find('h1')).toHaveLength(0);
    }); 
    it('should render h1 if we are not logged',()=>{
        localStorage.removeItem('token'); //why? 
        wrapper = mount( // wrapper update not working lol
            <Provider store={store}>
            <Social />
        </Provider>
        );
        expect(wrapper.find('h1').at(0).text()).toEqual('Login first 📜');
        expect(wrapper.find('h1').at(1).text()).toEqual("And Start Messaging ✏️ ");
        expect(wrapper.find(Button).text()).toEqual('Login');
    });
    it('should fire function when we go to social',()=>{
        const getUsers = jest.fn(); 
        const handleRender = jest.spyOn(React,'useEffect');
        handleRender.mockImplementation(user=>[user,getUser]);
        expect(getUsers).toBeTruthy();
    });
});