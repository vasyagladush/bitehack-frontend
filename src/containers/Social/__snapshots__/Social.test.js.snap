// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Social/> matches snapshot 1`] = `
Array [
  <div
    className="nav"
  >
    <p
      className="active"
    >
      Social
    </p>
    <p
      onClick={[Function]}
    >
      Home
    </p>
    <p
      onClick={[Function]}
    >
      About
    </p>
  </div>,
  <form
    className="socialinput"
  >
    <div
      className="form-group"
    >
      <input
        className="form-control"
        id="formBasicEmail"
        onChange={[Function]}
        placeholder="Search Friends by typing their names 🧙‍♂️"
        type="text"
        value=""
      />
    </div>
  </form>,
  <ul
    className="ul"
  >
    <h1
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      We can't find users 
      <span
        aria-hidden="true"
        aria-label=""
        className="emoji"
        role="img"
      >
        😵
      </span>
    </h1>
  </ul>,
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  />,
]
`;
